ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp

```diff
@@ -8747,3 +8904,323 @@ const GDALRelationship *GDALGeoPackageDataset::GetRelationship( const std::strin
+                                                                       relationship->GetRelatedTableType().c_str() );
+    // sanity checks
+    if( GetRelationship(osRelationshipName) != nullptr )
+    {
+        failureReason = "A relationship of identical name already exists";
+        return false;
+    }
+
+    if ( !ValidateRelationship(relationship.get(), failureReason ) )
+    {
+        return false;
+    }
+
+    if( CreateExtensionsTableIfNecessary() != OGRERR_NONE )
+    {
+        return false;
+    }
+    if ( !CreateRelationsTableIfNecessary() )
+    {
+        failureReason = "Could not create gpkgext_relations table";
+        return false;
+    }
+    if( SQLGetInteger(GetDB(),
+        "SELECT 1 FROM gpkg_extensions WHERE "
+        "table_name = 'gpkgext_relations'", nullptr) != 1 )
+    {
+        if( OGRERR_NONE != SQLCommand(GetDB(),
+            "INSERT INTO gpkg_extensions "
+            "(table_name,column_name,extension_name,definition,scope) "
+            "VALUES ('gpkgext_relations', NULL, 'gpkg_related_tables', "
+            "'http://www.geopackage.org/18-000.html', "
+            "'read-write')") )
+        {
+            failureReason = "Could not create gpkg_extensions entry for gpkgext_relations";
+            return false;
+        }
+    }
+
+    const std::string osLeftTableName = relationship->GetLeftTableName();
+    const std::string osRightTableName = relationship->GetRightTableName();
+    const auto& aosLeftTableFields = relationship->GetLeftTableFields();
+    const auto& aosRightTableFields = relationship->GetRightTableFields();
+
+    std::string osRelatedTableType = relationship->GetRelatedTableType();
+    if ( osRelatedTableType.empty() )
+    {
+        osRelatedTableType = "features";
+    }
+
+    // generate mapping table if not set
+    CPLString osMappingTableName = relationship->GetMappingTableName();
+    if ( osMappingTableName.empty() )
+    {
+        int nIndex = 1;
+        osMappingTableName = osLeftTableName + "_" + osRightTableName;
+        while ( FindLayerIndex( osMappingTableName.c_str() ) >= 0 )
+        {
+            nIndex += 1;
+            osMappingTableName.Printf( "%s_%s_%d", osLeftTableName.c_str(), osRightTableName.c_str(), nIndex );
+        }
+
+        // create mapping table
+        // note ideally we'd be setting foreign key constraints on this mapping table too, BUT this will cause sqlite
+        // foreign key mismatches if the base/related table fields aren't primary keys in their tables.
+        // "CONSTRAINT [fk_base_id_%q] FOREIGN KEY (base_id) REFERENCES %q(%q) ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED, "
+        // "CONSTRAINT [fk_related_id_%q] FOREIGN KEY (related_id) REFERENCES %q(%q) ON DELETE CASCADE ON UPDATE CASCADE DEFERRABLE INITIALLY DEFERRED"
+
+        char* pszSQL = sqlite3_mprintf("CREATE TABLE %q ("
```

Comment:
%q is for literals and must be surrounded with single quote. For identifiers you want %w surrounded with double quote
```suggestion
        char* pszSQL = sqlite3_mprintf("CREATE TABLE \"%w\" ("
```